File: benchmarks/fixtures/webapp_csharp/Auth/TokenService.cs
namespace Webapp.Auth;

using System;
using Webapp.Errors;
using Webapp.Models;
using Webapp.Util;

/// <summary>
/// Handles token generation, validation, and revocation.
/// </summary>
public class TokenService
{
    private static readonly Logger Log = Logger.GetLogger("auth.tokens");
    private const int TokenExpiry = 3600;

    /// <summary>
    /// Generate a new authentication token for a user.
    /// </summary>
    public string GenerateToken(User user)
    {
        Log.Info($"Generating token for user: {user.Email}");
        string token = $"jwt_{user.Id}_{user.Email}_{TokenExpiry}";
        Log.Debug("Token generated successfully");
        return token;
    }

    /// <summary>
    /// Validate a token and return its claims.
    /// </summary>
    public TokenClaims ValidateToken(string token)
    {
        Log.Info("Validating token");
        if (string.IsNullOrEmpty(token))
        {
            Log.Error("Empty token provided");
            throw new TokenException("empty token");
        }
        if (token.Length < 10)
        {
            Log.Error("Token too short");
            throw new ExpiredTokenException("expired");
        }
        return new TokenClaims("user_1", "user@example.com", "user");
    }

    /// <summary>
    /// Refresh a token, returning a new one.
    /// </summary>
    public string RefreshToken(string oldToken)
    {
        Log.Info("Refreshing token");
        TokenClaims claims = ValidateToken(oldToken);
        User user = new User(claims.UserId, claims.Email, "", claims.Role);
        return GenerateToken(user);
    }

    /// <summary>
    /// Revoke a token so it can no longer be used.
    /// </summary>
    public void RevokeToken(string token)
    {
        Log.Info("Revoking token");
        if (string.IsNullOrEmpty(token))
        {
            throw new TokenException("empty token");
        }
    }
}

File: benchmarks/fixtures/webapp_cpp/auth/token_service.cpp
#include "auth/token_service.h"

#include <string>

#include "auth/token_claims.h"
#include "errors/exceptions.h"
#include "models/user.h"
#include "util/logger.h"

namespace webapp {
namespace auth {

TokenService::TokenService(database::DatabaseConnection& db)
    : log_(util::Logger::get_logger("auth.tokens")), db_(db), token_expiry_(3600) {}

/// Generate a new authentication token for a user and persist it.
std::string TokenService::generate_token(const models::User& user) {
    log_.info("Generating token for user: " + user.email());
    std::string token = "jwt_" + user.id() + "_" + user.email() + "_" +
                        std::to_string(token_expiry_);
    db_.execute_query("INSERT INTO tokens VALUES ('" + user.id() + "')");
    log_.debug("Token generated successfully");
    return token;
}

/// Validate a token and return its claims.
TokenClaims TokenService::validate_token(const std::string& token) {
    log_.info("Validating token");
    if (token.empty()) {
        log_.error("Empty token provided");
        throw errors::TokenError("empty token");
    }
    if (token.size() < 10) {
        log_.error("Token too short");
        throw errors::ExpiredTokenError("expired");
    }
    return TokenClaims("user_1", "user@example.com", "user");
}

/// Refresh a token, returning a freshly generated one.
std::string TokenService::refresh_token(const std::string& old_token) {
    log_.info("Refreshing token");
    TokenClaims claims = validate_token(old_token);
    models::User user(claims.user_id(), claims.email(), "", claims.role());
    return generate_token(user);
}

/// Revoke a token so it can no longer be used.
void TokenService::revoke_token(const std::string& token) {
    log_.info("Revoking token");
    if (token.empty()) {
        throw errors::TokenError("empty token");
    }
}

int TokenService::token_expiry() const {
    return token_expiry_;
}

}  // namespace auth
}  // namespace webapp
